### ISSUE TITLE
Clear local DOM storage

### ISSUE THREAD

**Issue by ProgerXP** [dillbyrne/random-agent-spoofer#126]
Currently there's just one choice - to disable the storage. But it would make more sense to have it cleared on exit instead - just like Flash and normal cookies. Some sites (e.g. maps) require storage and rightfully so. Disabling it altogether isn't necessary as long as information there doesn't survive the restart. Basically, treat it as extended cookies.

**Comment by ARwMq9b6** [comment]
Yes it is, https://addons.mozilla.org/en-US/firefox/addon/cookie-controller/ will be helpful without dom disabled.

**Comment by ProgerXP** [comment]
Thanks for the link, I'll use this addon for now. However, it would be nice to have in RAS core since it already can disable the storage.

**Comment by CHEF-KOCH** [comment]
Since this would break sites which depending on it like Mega.co.cz there should also be an whitelist option. Currently Firefox has only an enabled/disabled statement and any other stuff needs an extra addon, that really sux.

**Comment by ProgerXP** [comment]
> Since this would break sites which depending on it like Mega.co.cz there should also be an whitelist option. 

There should be an option to clear the storage. But in fact I have solved this problem already. There are two ways:
* Use this addon that does exactly that - clear LS on startup and/or shutdown: https://addons.mozilla.org/en-US/firefox/addon/volatile-storage/
* Or set Firefox Options > Privacy > Custom settings > Clear cookies until I close Firefox - this surprisingly has the effect of also clearing LS when Firefox exits

So I think RAS can just ignore this feature request as there is even a standard way to handle it.

**Comment by CHEF-KOCH** [comment]
Well, I already did say that using another addon sux, so I don't know why you suggest that now, thanks anyway. 
Nope, _clear cookies until I close Firefox_ does not delete any dom storage since this is more like a supercookie (especially with the Mega addon) and is _harder_ to kill on sites like Mega (which stores the key in there and other important things..) and as I said an whitelist option would be very nice to avoid disabling the entire option and just use the specific sites people really needs.

```
dom.storage.enabled
dom.storage.default_quota
```

Since FF 34 there is a [StorageInspector](https://developer.mozilla.org/en-US/docs/Tools/Storage_Inspector) which allow us to clear that via:
* DOM Storage can be cleared via "Tools -> Clear Recent History -> Cookies" when Time range is "Everything" (via nsICookieManager::removeAll) but that only works if the cookie supports it
* Via console _localStorage.clear();_ - which always works
* See also [Mozilla's DOM Storage Guide](https://developer.mozilla.org/en-US/docs/Web/Guide/API/DOM/Storage) under _Storage location and clearing the data_

So an easier option to do that via RAS would be much welcome here.

**Comment by ProgerXP** [comment]
> Well, I already did say that using another addon sux, so I don't know why you suggest that now, thanks anyway. 

Because unlike other addons (e.g. suggested Cookie Controller) Volatile Storage does only one thing and its impact on anything from performance to UI is negligible.

> Nope, clear cookies until I close Firefox does not delete any dom storage since this is more like a supercookie 

Have you tried it yourself? Simple experiment shows that it does, for me. First set that setting, then open the console on any site that you haven't whitelisted cookies on, run `localStorage.setItem('abc',123)`, restart Firefox, go to that site again and you will see that `localStorage` contains 0 elements. On the other hand, if you Allow cookies for a particular site and store any item in LS on that domain - it will persist after restart.

Then again if you set an LS item and then check the contents of `webappsstore.sqlite` you will see that no items appear in there when they are set.

**Comment by dillbyrne** [comment]
Hello everyone .

Sorry for the late reply. I was injured and unable to type. 

This is on the road map and I will try to get a working solution in the next release or two. I have not had a proper chance to try it out. It will have to do done using script injection, this will allow us to selectively whitelist it and block it elsewhere.

It should be sufficient to block the setter and getter methods of the `localStorage` object and make them return null or undefined if they are not whitelisted.
